google/cloud/aiplatform/docker_utils/build.py

```diff
@@ -262,35 +297,65 @@ def make_dockerfile(
         """
         FROM {base_image}
 
+        # Keeps Python from generating .pyc files in the container
+        ENV PYTHONDONTWRITEBYTECODE=1
+        """.format(
+            base_image=base_image,
+        )
+    )
+
+    dockerfile += _prepare_exposed_ports(exposed_ports)
+
+    dockerfile += _prepare_entrypoint(main_package, python_command=python_command)
+
+    dockerfile += textwrap.dedent(
+        """
         # The directory is created by root. This sets permissions so that any user can
         # access the folder.
         RUN mkdir -m 777 -p {workdir} {container_home}
         WORKDIR {workdir}
         ENV HOME={container_home}
-
-        # Keeps Python from generating .pyc files in the container
-        ENV PYTHONDONTWRITEBYTECODE=1
         """.format(
-            base_image=base_image,
             workdir=shlex_quote(container_workdir),
             container_home=shlex_quote(container_home),
         )
     )
 
-    dockerfile += _prepare_package_entry(main_package)
+    # Installs extra requirements which do not involve user source code.
+    dockerfile += _prepare_dependency_entries(
```

Comment:
Why do the requirements_path, extra_requirements, etc, need to be separated into individual _prepare_dependency_entries calls?

Reply:
Separating the calls allows us to control the order of the operations in Dockerfile instead of depending on the code in `_prepare_dependency_entries`. Therefore, we could install the things that are likely to be unchanged first to optimize the time building images with docker build cache.

Reply:
Ah, got it. Thanks.